gateway/gateway.go

```diff
@@ -17,34 +17,122 @@ type Config struct {
 	Headers map[string][]string
 }
 
-// API defines the minimal set of API services required for a gateway handler.
-type API interface {
-	// GetUnixFsNode returns a read-only handle to a file tree referenced by a path.
-	GetUnixFsNode(context.Context, path.Resolved) (files.Node, error)
+// TODO: Is this what we want for ImmutablePath?
+type ImmutablePath struct {
+	p path.Path
+}
+
+func NewImmutablePath(p path.Path) (ImmutablePath, error) {
+	if p.Mutable() {
+		return ImmutablePath{}, fmt.Errorf("path cannot be mutable")
+	}
+	return ImmutablePath{p: p}, nil
+}
+
+func (i ImmutablePath) String() string {
+	return i.p.String()
+}
+
+func (i ImmutablePath) Namespace() string {
+	return i.p.Namespace()
+}
+
+func (i ImmutablePath) Mutable() bool {
+	return false
+}
+
+func (i ImmutablePath) IsValid() error {
+	return i.p.IsValid()
+}
+
+var _ path.Path = (*ImmutablePath)(nil)
+
+type GatewayMetadata struct {
+	PathSegmentRoots []cid.Cid
+	LastSegment      path.Resolved
+}
+
+// TODO: These functional options seems a little unwieldly here and require a bunch of text, would just having more functions be better?
+type GetOpt func(*GetOptions) error
```

Comment:
@hacdias @Jorropo I think you wanted options here, but it's a bit unfriendly how we make every consumer deal with them and how we have to add more (e.g. text or code) around the API to make sure people don't use multiple options when they only should support one

Reply:
I'm not sure what you mean by "wanted options" - I don't think I've mentioned this before. Maybe it was @Jorropo?

Reply:
It was you https://www.notion.so/WIP-go-libipfs-Gateway-GO-API-6f6dc875e65544e58740fadd206b2e74?d=77f2539067804b75977cf0d9f2fd7ac4#1ccd07ef4f4943d1b3670c11500b33da, but you don't seem to feel too strongly so going back to the original draft 😄

Reply:
Ahhh, I see what you mean by options then. Yes, I don't feel that strongly about any of the options. Separate functions are fine for me too.